compiler/src/Language/Bond/Codegen/Java/Class_java.hs

```diff
@@ -373,6 +373,41 @@ public #{declName}(com.microsoft.bond.StructBondType<#{declName}<#{paramList}>>
             else mempty
 
 
+-- builds text for Object.equals(Object) override
+object_equals :: String -> [Field] -> Text
+object_equals declName fields = [lt|@Override
+    public boolean equals(Object o) {
+        if (o == null) return false;
+        if (!(o instanceof #{declName})) return false;
+        final #{declName} other = (#{declName}) o;
+        #{newlineSep 2 compareField fields}
+        return true;
+    }
+|]
+    where
+        compareField Field {..} = [lt|if (!(#{equals_member fieldType fieldName})) return false;|]
+        equals_member t f
+            | isJavaPrimitive t = [lt|this.#{f} == other.#{f}|]
```

Comment:
NaN == NaN is always false. Further, 0D == -0D is always true. There is FloatingPointHelper class that compares bitwise, which is what we use when serializing.

Java has different comparison implementations for primitive double and object Double types; the latter is comparing bits. We should keep this consistent and compare primitives using bits. Note that collections already do the right comparison since elements are objects.

---
In reply to: [131460893](https://github.com/Microsoft/bond/pull/545#discussion_r131460893) [](ancestors = 131460893)